internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,604 @@
+			svc.Spec.Ports = append(svc.Spec.Ports, corev1.ServicePort{})
+		}
+		// For convenience targetPort, when unset, is set to the same as port
+		// https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service
+		// so we don't set it.
+		svc.Spec.Ports[0].Name = name
+		svc.Spec.Ports[0].Port = servicePort
+		svc.Spec.Ports[0].Protocol = corev1.ProtocolTCP
+		svc.Spec.Selector = map[string]string{"app": name}
+		svc.Spec.Type = corev1.ServiceTypeLoadBalancer
+		return nil
+	}
+
+	return svc, mutateFn
+}
+
+func (r *AgentServiceConfigReconciler) newAgentRoute(instance *adiiov1alpha1.AgentServiceConfig) (*routev1.Route, controllerutil.MutateFn) {
+	weight := int32(100)
+	route := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.Namespace,
+		},
+	}
+	routeSpec := routev1.RouteSpec{
+		To: routev1.RouteTargetReference{
+			Kind:   "Service",
+			Name:   name,
+			Weight: &weight,
+		},
+		Port: &routev1.RoutePort{
+			TargetPort: intstr.FromString(name),
+		},
+		WildcardPolicy: routev1.WildcardPolicyNone,
+	}
+
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(instance, route, r.Scheme); err != nil {
+			return err
+		}
+		route.Spec = routeSpec
+		return nil
+	}
+
+	return route, mutateFn
+}
+
+func (r *AgentServiceConfigReconciler) newPostgresSecret(instance *adiiov1alpha1.AgentServiceConfig) (*corev1.Secret, controllerutil.MutateFn) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      databaseSecretName,
+			Namespace: r.Namespace,
+		},
+	}
+
+	// pass, err := generatePassword()
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	// TODO(djzager): Use legit password
+	secretData := map[string]string{
+		"db.host":     "localhost",
+		"db.user":     "admin",
+		"db.password": "abcdefg",
+		"db.name":     "installer",
+		"db.port":     strconv.Itoa(int(databasePort)),
+	}
+
+	secretType := corev1.SecretTypeOpaque
+
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(instance, secret, r.Scheme); err != nil {
+			return err
+		}
+		secret.StringData = secretData
+		secret.Type = secretType
+		return nil
+	}
+
+	return secret, mutateFn
+}
+
+func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(instance *adiiov1alpha1.AgentServiceConfig) (*appsv1.Deployment, controllerutil.MutateFn) {
+	ocmSSOSecretName := "assisted-installer-sso"
+	s3SecretName := "assisted-installer-s3"
+	publicS3SecretName := "assisted-installer-public-s3"
+	assistedServiceConfigMapName := "assisted-service-config"
+	maxUnavailable := intstr.FromString("50%")
+	maxSurge := intstr.FromString("100%")
+
+	serviceContainer := corev1.Container{
+		Name:  assistedServiceContainerName,
+		Image: ServiceImage(),
+		Ports: []corev1.ContainerPort{
+			{
+				ContainerPort: servicePort,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		},
+		Env: []corev1.EnvVar{
+			// TODO(djzager): clean up unneccessary environment variables
```

Comment:
If we set the env vars separately from the container in the mutate function we could support something like @filanov wanted by only ensuring the values we want are present rather than setting the whole array.

Reply:
I don't think that we need to set everything here.
Most of the things have defaults in the code, we don't need to duplicate it.

Reply:
I think that is a reasonable way to approach it. My personal opinion though is that, in the short term, we should avoid letting the user configure anything that we don't explicitly want to support.

But I'm going to create a Jira to track it and then we can prioritize it.